lib/core/include/shared_memory_object.hpp

```diff
@@ -0,0 +1,74 @@
+#ifndef IRODS_SHARED_MEMORY_OBJECT_HPP
+#define IRODS_SHARED_MEMORY_OBJECT_HPP
+
+#include <boost/interprocess/shared_memory_object.hpp>
+#include <boost/interprocess/mapped_region.hpp>
+#include <boost/interprocess/sync/interprocess_mutex.hpp>
+#include <boost/interprocess/sync/scoped_lock.hpp>
+
+#include <string>
+#include <utility>
+
+namespace irods::experimental::interprocess
+{
+    template <typename T>
+    class shared_memory_object
+    {
+    private:
+        struct ipc_object
+        {
+            T thing;
+            boost::interprocess::interprocess_mutex mtx;
+        };
+
+    public:
+        template <typename ...Args>
+        shared_memory_object(std::string _shm_name, Args&& ..._args)
+            : shm_name_{std::move(_shm_name)}
+            , shm_{boost::interprocess::open_or_create, shm_name_.c_str(), boost::interprocess::read_write}
+            , region_{}
+            , object_{}
+        {
+            boost::interprocess::offset_t size;
+
+            // The shared memory was created if the size is equal to zero.
+            if (shm_.get_size(size) && 0 == size) {
+                shm_.truncate(sizeof(ipc_object));
+                region_ = {shm_, boost::interprocess::read_write};
+                object_ = new (region_.get_address()) ipc_object{T{std::forward<Args>(_args)...}, {}};
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Class 'shared_memory_object' does not have a copy constructor which is recommended since it has dynamic memory/resource allocation(s).](https://app.codacy.com/manual/trel/irods/pullRequest?prid=5130675)